internal/topology/upgrade/clusterctl_upgrade_test.go

```diff
@@ -0,0 +1,1174 @@
+	}
+	if c := conditions.Get(actualCluster, clusterv1.ClusterTopologyReconciledCondition); c != nil && c.Status != metav1.ConditionTrue {
+		return nil, errors.Errorf("cluster %s topology is not reconciled", cluster.Name)
+	}
+
+	addToAllObj(cluster.Name, "cluster", actualCluster)
+	refs[clusterv1.ContractVersionedObjectReference{
+		APIGroup: actualCluster.GroupVersionKind().Group,
+		Kind:     actualCluster.GroupVersionKind().Kind,
+		Name:     actualCluster.GetName(),
+	}] = actualCluster.GetGeneration()
+
+	if actualCluster.Spec.InfrastructureRef == nil {
+		return nil, errors.New("cluster actualCluster.spec.infrastructureRef is not yet set")
+	}
+	refObj, err := getReferencedObject(ctx, env.GetClient(), actualCluster.Spec.InfrastructureRef, version, actualCluster.Namespace)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get referenced actualCluster.spec.infrastructureRef")
+	}
+	addToAllObj(cluster.Name, ".spec.infrastructureRef", refObj)
+	refs[clusterv1.ContractVersionedObjectReference{
+		APIGroup: refObj.GroupVersionKind().Group,
+		Kind:     refObj.GetKind(),
+		Name:     refObj.GetName(),
+	}] = refObj.GetGeneration()
+	if additionalChecks != nil {
+		if err := additionalChecks(refObj); err != nil {
+			return nil, errors.Wrap(err, "failed additional checks on actualCluster.spec.infrastructureRef")
+		}
+	}
+
+	if actualCluster.Spec.ControlPlaneRef == nil {
+		return nil, errors.New("cluster actualCluster.spec.controlPlaneRef is not yet set")
+	}
+	refObj, err = getReferencedObject(ctx, env.GetClient(), actualCluster.Spec.ControlPlaneRef, version, actualCluster.Namespace)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get referenced actualCluster.spec.controlPlaneRef")
+	}
+	addToAllObj(cluster.Name, ".spec.controlPlaneRef", refObj)
+	refs[clusterv1.ContractVersionedObjectReference{
+		APIGroup: refObj.GroupVersionKind().Group,
+		Kind:     refObj.GetKind(),
+		Name:     refObj.GetName(),
+	}] = refObj.GetGeneration()
+	if additionalChecks != nil {
+		if err := additionalChecks(refObj); err != nil {
+			return nil, errors.Wrap(err, "failed additional checks on actualCluster.spec.controlPlaneRef")
+		}
+	}
+
+	cpInfraRef, err := contract.ControlPlane().MachineTemplate().InfrastructureRef().Get(refObj)
+	if err != nil {
+		return nil, errors.Wrap(err, "cluster controplPlane.spec.machineTemplate.infrastructureRef is not yet set")
+	}
+	refObj, err = getReferencedObject(ctx, env.GetClient(), cpInfraRef, version, actualCluster.Namespace)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get referenced controplPlane.spec.machineTemplate.infrastructureRef")
+	}
+	addToAllObj(cluster.Name, "controlPlane.spec.machineTemplate.infrastructureRef", refObj)
+	refs[clusterv1.ContractVersionedObjectReference{
+		APIGroup: refObj.GroupVersionKind().Group,
+		Kind:     refObj.GetKind(),
+		Name:     refObj.GetName(),
+	}] = refObj.GetGeneration()
```

Comment:
Should we also run additionalChecks here? (please check that we run it in all places where we want to run them)